pkg/postage/postagecontract/contract.go

```diff
@@ -0,0 +1,187 @@
+	postageStampAddress common.Address
+	bzzTokenAddress     common.Address
+	transactionService  transaction.Service
+	postageService      postage.Service
+}
+
+func New(
+	owner common.Address,
+	postageStampAddress common.Address,
+	bzzTokenAddress common.Address,
+	transactionService transaction.Service,
+	postageService postage.Service,
+) Interface {
+	return &postageContract{
+		owner:               owner,
+		postageStampAddress: postageStampAddress,
+		bzzTokenAddress:     bzzTokenAddress,
+		transactionService:  transactionService,
+		postageService:      postageService,
+	}
+}
+
+func (c *postageContract) sendApproveTransaction(ctx context.Context, amount *big.Int) (*types.Receipt, error) {
+	callData, err := erc20ABI.Pack("approve", c.postageStampAddress, amount)
+	if err != nil {
+		return nil, err
+	}
+
+	txHash, err := c.transactionService.Send(ctx, &transaction.TxRequest{
+		To:       c.bzzTokenAddress,
+		Data:     callData,
+		GasPrice: nil,
+		GasLimit: 0,
+		Value:    big.NewInt(0),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := c.transactionService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if receipt.Status == 0 {
+		return nil, transaction.ErrTransactionReverted
+	}
+
+	return receipt, nil
+}
+
+func (c *postageContract) sendCreateBatchTransaction(ctx context.Context, owner common.Address, initialBalance *big.Int, depth uint8, nonce common.Hash) (*types.Receipt, error) {
+	callData, err := postageStampABI.Pack("createBatch", owner, initialBalance, depth, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &transaction.TxRequest{
+		To:       c.postageStampAddress,
+		Data:     callData,
+		GasPrice: nil,
+		GasLimit: 0,
+		Value:    big.NewInt(0),
+	}
+
+	txHash, err := c.transactionService.Send(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := c.transactionService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if receipt.Status == 0 {
+		return nil, transaction.ErrTransactionReverted
+	}
+
+	return receipt, nil
+}
+
+func (c *postageContract) CreateBatch(ctx context.Context, initialBalance *big.Int, depth uint8) ([]byte, error) {
+	_, err := c.sendApproveTransaction(ctx, big.NewInt(0).Mul(initialBalance, big.NewInt(int64(1<<depth))))
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, 32)
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := c.sendCreateBatchTransaction(ctx, c.owner, initialBalance, depth, common.BytesToHash(nonce))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ev := range receipt.Logs {
+		if ev.Address == c.postageStampAddress && ev.Topics[0] == batchCreatedTopic {
+			var createdEvent batchCreatedEvent
+			err = parseEvent(&postageStampABI, "BatchCreated", &createdEvent, *ev)
+			if err != nil {
+				return nil, err
+			}
+
+			batchID := createdEvent.BatchId[:]
+
+			c.postageService.Add(postage.NewStampIssuer(
+				"label",
```

Comment:
@zelig please elaborate what these are for and if we should even set them now

Reply:
yes, please set them via query param and/or form-data .
They are meeant to support 2 dimensional classification of stamps. 
The UX presented in the book can then be simply supported from js side, 
https://gateway.ethswarm.org/bzz/latest.bookofswarm.eth/#subsection.6.1.3 
with `label` being generated from:

```
8\minimal (a few hours)– Useful for immediate delivery of pss messages or singleowner chunks part of ephemeral chat or other temporary files.
•temporary (week)– Files or mailboxed messages meant not to be stored long butto be picked up by third parties asynchronously.
•long term (year)– Default long term storage.
•forever (10years)– Important content not to be lost/forgotten;  to survive thegrowth  of  the  network  and  subsequent  increase  of  batch  depth  even  if  theuploader remains completely offline.
```

and batch reuse strategy maybe

"keyid" can be recorded as the batch owner etheraddress that would be used to call the appropriate signer with the right batch owner key.

Reply:
`label` query param added. `keyID` is hex representation of owner address.